sync15-adapter/src/token.rs

```diff
@@ -131,3 +143,185 @@ impl TokenserverClient {
+                }
+            },
+            TokenState::NodeReassigned => {
+                // We never leave this state.
+                None
+            }
+        }
+    }
+
+    fn with_token<T, F>(&self, request_client: &Client, func: F) -> Result<T>
+            where F: FnOnce(&TokenContext) -> Result<T> {
+
+        // first get a mutable ref to our existing state, advance to the
+        // state we will use, then re-stash that state for next time.
+        let state: &mut TokenState = &mut self.current_state.borrow_mut();
+        match self.advance_state(request_client, state) {
+            Some(new_state) => *state = new_state,
+            None => ()
+        }
+
+        // Now re-fetch the state we should use for this call - if it's
+        // anything other than TokenState::Token we will fail.
+        match state {
+            TokenState::NoToken => {
+                // it should be impossible to get here.
+                panic!("Can't be in NoToken state after advancing");
+            }
+            TokenState::Token(ref token_context) => {
+                // make the call.
+                func(token_context)
+            }
+            TokenState::Failed(_, _) => {
+                // XXX - todo - work out how to return the value???
+                bail!(error::unexpected("todo - work out how to return the actual error?"))
+            }
+            TokenState::NodeReassigned => {
+                // this is unrecoverable.
+                bail!(error::ErrorKind::StorageResetError);
+            }
+            TokenState::Backoff(ref remaining, _) => {
+                bail!(error::ErrorKind::BackoffError(*remaining));
+            }
+        }
+    }
+
+    pub fn authorization(&self, http_client: &Client, req: &Request) -> Result<Authorization<String>> {
+        Ok(self.with_token(http_client, |ctx| ctx.authorization(req))?)
+    }
+
+    pub fn api_endpoint(&self, http_client: &Client) -> Result<String> {
+        Ok(self.with_token(http_client, |ctx| Ok(ctx.token.api_endpoint.clone()))?)
+    }
+    // TODO: we probably want a "drop_token/context" type method so that when
+    // using a token with some validity fails the caller can force a new one
+    // (in which case the new token request will probably fail with a 401)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use reqwest::{Client};
+
+    #[test]
+    fn test_something() {
+        let client = Client::builder().timeout(Duration::from_secs(30)).build().expect("can't build client");
+        let tsc = TokenserverClient::new(String::from("base_url"),
+                                         String::from("access_token"),
+                                         String::from("key_id"));
+
+        // TODO: make this actually useful!
+        let _e = tsc.api_endpoint(&client).expect_err("should fail");
+        // XXX - (a) it's an unexpected error rather than the original, and
```

Comment:
I think if we care what it says (even in tests) we shouldn't use `UnexpectedError` for it.